gbus/tx/mysql/timeout.go

```diff
@@ -0,0 +1,203 @@
+package mysql
+
+import (
+	"database/sql"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/sirupsen/logrus"
+
+	"github.com/wework/grabbit/gbus"
+)
+
+var _ gbus.TimeoutManager = &TimeoutManager{}
+
+//TimeoutManager is a mysql implementation of a persistent timeoutmanager
+type TimeoutManager struct {
+	Bus         gbus.Bus
+	Log         func() logrus.FieldLogger
+	TimeoutSaga func(*sql.Tx, string) error
+	Txp         gbus.TxProvider
+	SvcName     string
+	exit        chan bool
+}
+
+func (tm *TimeoutManager) ensureSchema() error {
+	tblName := tm.GetTimeoutsTableName()
+	tx, e := tm.Txp.New()
+	if e != nil {
+		tm.Log().WithError(e).Error("failed to create schema for mysql timeout manager")
+		return e
+	}
+
+	createTableSQL := `CREATE TABLE IF NOT EXISTS ` + tblName + ` (
+      rec_id INT PRIMARY KEY AUTO_INCREMENT,
+      saga_id VARCHAR(255) UNIQUE NOT NULL,
+	  timeout DATETIME NOT NULL,
+	  INDEX ix_` + tm.GetTimeoutsTableName() + `_timeout_date(timeout)
+      )`
+
+	if _, e := tx.Exec(createTableSQL); e != nil {
+		if rbkErr := tx.Rollback(); rbkErr != nil {
+			tm.Log().Warn("timeout manager failed to rollback transaction")
+		}
+		return e
+	}
+	return tx.Commit()
+}
+
+func (tm *TimeoutManager) purge() error {
+	purgeSQL := `DELETE FROM ` + tm.GetTimeoutsTableName()
+
+	tx, e := tm.Txp.New()
+	if e != nil {
+		tm.Log().WithError(e).Error("failed to purge timeout manager")
+		return e
+	}
+
+	if _, execErr := tx.Exec(purgeSQL); execErr != nil {
+		tm.Log().WithError(execErr).Error("failed to purge timeout manager")
+		return tx.Rollback()
+	}
+	return tx.Commit()
+}
+
+//Start starts the timeout manager
+func (tm *TimeoutManager) Start() error {
+	go tm.trackTimeouts()
+	return nil
+}
+
+//Stop shuts down the timeout manager
+func (tm *TimeoutManager) Stop() error {
+	tm.exit <- true
+	return nil
+}
+
+func (tm *TimeoutManager) trackTimeouts() {
+	tick := time.NewTicker(time.Second * 1).C
+	for {
+		select {
+		case <-tick:
+			tx, txe := tm.Txp.New()
+			if txe != nil {
+				tm.Log().WithError(txe).Warn("timeout manager failed to create a transaction")
+				continue
+			}
+			now := time.Now().UTC()
+			getTimeoutsSQL := `select saga_id from ` + tm.GetTimeoutsTableName() + ` where timeout < ? LIMIT 100`
```

Comment:
```suggestion
			getTimeoutsSQL := `select saga_id from ` + tm.GetTimeoutsTableName() + ` where timeout < ? FOR UPDATE SKIP LOCKED LIMIT 100`
```
This way if you have multiple instances runnnig they will not process the same sagas.

Reply:
@vladshub this is actually not the correct place to lock the record as I want each timeout to run in its own transaction so can't take and hold a lock here.
Will add locking the timeout record in the executeTimeouts function to ensure a timeout will only get executed once in the scenario you mentioned

Reply:
I have placed a "FOR UPDATE" lock in the scope of the specific transaction running the timeout before executing the timeout.
If the lock can not be acquired the timeout is skipped and transaction rolledback